### PR TITLE
ccacheStdenv: support cross compilation

### PR DESCRIPTION
**Pull request by chayleaf**
## Description of changes

To test:

1. Set `programs.ccache.enable` to `true` in system config (or just create `/var/cache/ccache` owned by group `nixbld` with mode 770)

2.

```nix
pkgsCross.aarch64-multiplatform.hello.override {
  stdenv = pkgsCross.aarch64-multiplatform.ccacheStdenv.override {
    extraConfig = ''      
      export CCACHE_DIR=/var/cache/ccache
      export CCACHE_UMASK=007
    '';
  };
}
```

3. Make sure `/var/cache/ccache` contains the cached artifacts

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ConnorBaker** [comment]
Pinging maintainers @kira-bruneau and @r-burns for review

**Comment by r-burns** [comment]
Thanks, weird that ofborg didn't ping me for this, I guess because no rebuilds? Don't have time to test at the moment but changes look good to me.

File: pkgs/development/tools/misc/ccache/default.nix [hunk 1]
```diff
@@ -98,12 +98,12 @@ stdenv.mkDerivation (finalAttrs: {
...
-        wrap gcc
-        wrap g++
-        wrap clang
-        wrap clang++
+        wrap ${binutils.targetPrefix}cc
+        wrap ${binutils.targetPrefix}c++
+        wrap ${binutils.targetPrefix}gcc
+        wrap ${binutils.targetPrefix}g++
+        wrap ${binutils.targetPrefix}clang
+        wrap ${binutils.targetPrefix}clang++
```
Comment (Artturin) [inline pkgs/development/tools/misc/ccache/default.nix:106]:
```suggestion
        wrap ${stdenv.cc.targetPrefix}cc
        wrap ${stdenv.cc.targetPrefix}c++
        wrap ${stdenv.cc.targetPrefix}gcc
        wrap ${stdenv.cc.targetPrefix}g++
        wrap ${stdenv.cc.targetPrefix}clang
        wrap ${stdenv.cc.targetPrefix}clang++
```

Comment (Artturin) [reply pkgs/development/tools/misc/ccache/default.nix:106]: well this is not really correct either

Comment (Artturin) [reply pkgs/development/tools/misc/ccache/default.nix:106]:
```suggestion
        wrap ${lib.optionalString unwrappedCC.targetConfig "${unwrappedCC.targetConfig}-"}cc
        wrap ${lib.optionalString unwrappedCC.targetConfig "${unwrappedCC.targetConfig}-"}c++
        wrap ${lib.optionalString unwrappedCC.targetConfig "${unwrappedCC.targetConfig}-"}gcc
        wrap ${lib.optionalString unwrappedCC.targetConfig "${unwrappedCC.targetConfig}-"}g++
        wrap ${lib.optionalString unwrappedCC.targetConfig "${unwrappedCC.targetConfig}-"}clang
        wrap ${lib.optionalString unwrappedCC.targetConfig "${unwrappedCC.targetConfig}-"}clang++
```
Should be something like this
Put it in a let

**Review by Artturin — APPROVED** [review]

File: pkgs/development/tools/misc/ccache/default.nix [hunk 2]
```diff
@@ -86,7 +86,10 @@ stdenv.mkDerivation (finalAttrs: {
       };
       inherit (unwrappedCC) lib;
       nativeBuildInputs = [ makeWrapper ];
-      buildCommand = ''
+      buildCommand = let
+        # Clang does not yet have targetConfig in passthru.
+        targetPrefix = if stdenv.cc.isGNU then (lib.optionalString (unwrappedCC.targetConfig != null) "${unwrappedCC.targetConfig}-") else stdenv.cc.targetPrefix;
```
Comment (Artturin) [inline pkgs/development/tools/misc/ccache/default.nix:91]: Unwrapped clang does not have prefixes
`pkgsCross.aarch64-multiplatform.clangStdenv.cc.cc` 

clang uses argv0 to pick the target

**Comment by Artturin** [comment]
Hopefully now it works for gcc and clang

@chayleaf can you test

**Pull request merged**